epuck_ros2_camera/src/camera.cpp

```diff
@@ -87,63 +116,122 @@ class CameraPublisher : public rclcpp::Node {
+      mPublisherRaw->publish(message);
+    } else
+      deinitRgb();
 
+    // Publish JPEG compressed image if needed
     if (mPublisherCompressed->get_subscription_count() > 0) {
-      if (!mJpegInitialized) {
-        pipuck_jpeg_init(&mCapturedImage, &mCompressedImage);
-        mJpegInitialized = true;
-      }
-      pipuck_jpeg_encode(&mCapturedImage, &mCompressedImage);
+      initJpeg();
 
+      // It's the same data for both topics
+      mPipuckMmalJpeg.input.data = mPipuckMmalRgb.input.data;
+      mPipuckMmalJpeg.input.size = mPipuckMmalRgb.input.size;
+
+      pipuck_mmal_convert(&mPipuckMmalJpeg);
       auto message = sensor_msgs::msg::CompressedImage();
       message.format = "jpeg";
-      message.header.stamp = now();
+      message.header.stamp = stamp;
       message.header.frame_id = "pipuck_image_compressed";
-      message.data.assign(mCompressedImage.data, mCompressedImage.data + mCompressedImage.size);
-
+      message.data.assign(mPipuckMmalJpeg.output.data, mPipuckMmalJpeg.output.data + mPipuckMmalJpeg.output.size);
       mPublisherCompressed->publish(message);
-    } else if (mJpegInitialized) {
-      pipuck_jpeg_deinit();
+    } else
+      deinitJpeg();
+  }
+
+  void initV4l2() {
+    if (!mV4l2Initialized) {
+      pipuck_v4l2_init();
+      mV4l2Initialized = true;
+      RCLCPP_INFO(this->get_logger(), "V4L2 component initialized");
+    }
+  }
+
+  void deinitV4l2() {
+    if (mV4l2Initialized) {
+      pipuck_v4l2_deinit();
+      mV4l2Initialized = false;
+      RCLCPP_INFO(this->get_logger(), "V4L2 component deinitialized");
+      deinitJpeg();
+      deinitRgb();
+    }
+  }
+
+  void deinitJpeg() {
+    if (mJpegInitialized) {
+      pipuck_mmal_deinit(&mPipuckMmalJpeg);
       mJpegInitialized = false;
+      RCLCPP_INFO(this->get_logger(), "MMAL JPEG component deinitialized");
+    }
+  }
+
+  void initJpeg() {
+    if (!mJpegInitialized) {
+      pipuck_mmal_init(&mPipuckMmalJpeg);
+      mJpegInitialized = true;
+      RCLCPP_INFO(this->get_logger(), "MMAL JPEG component initialized");
+    }
+  }
+
+  void deinitRgb() {
+    if (mRgbInitialized) {
+      pipuck_mmal_deinit(&mPipuckMmalRgb);
+      mRgbInitialized = false;
+      RCLCPP_INFO(this->get_logger(), "MMAL RGB component deinitialized");
     }
   }
+
+  void initRgb() {
+    if (!mRgbInitialized) {
+      pipuck_mmal_init(&mPipuckMmalRgb);
+      mRgbInitialized = true;
+      RCLCPP_INFO(this->get_logger(), "MMAL RGB component initialized");
+    }
+  }
+
   rclcpp::TimerBase::SharedPtr mTimer;
+  rclcpp::Publisher<sensor_msgs::msg::CameraInfo>::SharedPtr mPublisherCameraInfo;
   rclcpp::Publisher<sensor_msgs::msg::CompressedImage>::SharedPtr mPublisherCompressed;
   rclcpp::Publisher<sensor_msgs::msg::Image>::SharedPtr mPublisherRaw;
-  pipuck_image_t mCapturedImage;
-  pipuck_image_t mCompressedImage;
   OnSetParametersCallbackHandle::SharedPtr mCallbackHandler;
   char mImageBuffer[900 * 1024];
   bool mV4l2Initialized;
```

Comment:
```suggestion
  bool mIsV4l2Initialized;
```

This is slightly more self-explanatory: https://github.com/cyberbotics/webots/wiki/CPP-Coding-Style#cs320-the-is-prefix-should-be-used-for-boolean-variables-and-methods